crates/wasmparser/src/validator/component.rs

```diff
@@ -952,7 +953,11 @@ impl ComponentState {
 
         self.check_options(None, &info, &options, types, offset)?;
 
-        let lowered_ty = Type::Func(info.into_func_type());
+        let lowered_ty = Type::Sub(SubType {
+            is_final: false,
+            supertype_idxs: vec![],
+            structural_type: StructuralType::Func(info.into_func_type()),
+        });
```

Comment:
Looking at all these conversions between `Type`, `SubType`, `StructuralType`, and each structural type variant, it seems like we could have some more `From` implementations to make this less noisy and more automatic.

To really benefit from this, and because `From` is not transitive by default, I think we would want all of these impls:

```rust
impl From<FuncType> for StructuralType { ... }
impl From<ArrayType> for StructuralType { ... }
impl From<StructType> for StructuralType { ... }

impl From<StructuralType> for SubType { ... }
impl From<FuncType> for SubType { ... }
impl From<ArrayType> for SubType { ... }
impl From<StructType> for SubType { ... }

impl From<SubType> for Type { ... }
impl From<StructuralType> for Type { ... }
impl From<FuncType> for Type { ... }
impl From<ArrayType> for Type { ... }
impl From<StructType> for Type { ... }
```

This can be done in a follow up PR if you prefer.

Reply:
Added an issue to follow this up: #1071